### PR TITLE
 TrimUnreferencedReaders: move sync translog operation outside writeLock to improve performance

### LINKED ISSUE #46201: Optimize translog writing by move sync outside writeLock in trimUnreferencedReaders
In Translog.trimUnreferencedReaders, current.sync() is called inside the writeLock.
current.sync() will take about one seconds as discuss in PR #45450 

So take it out of the writeLock can improve the indexing performance as below

**accumulated time of all index operation**: I accumulate the time took by every index operation

the benchmark is done in default config. about performance improve about 8%

### PR DESCRIPTION
**Pull request by dengweisysu**
close #46201

### CODE REVIEW

**Comment by dengweisysu** [comment]
@dnhatn benchmark is provide in #46201

**Comment by dnhatn** [comment]
@elasticmachine ok to test

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 1]
```diff
@@ -1707,6 +1701,24 @@ public void trimUnreferencedReaders() throws IOException {
             closeOnTragicEvent(ex);
             throw ex;
         }
+        // Do sync outside the writeLock to avoid blocking all writing thread.
+        try {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java]: It's easier to follow with `try-catch` inside `if(toDeleteReaderList.isEmpty() == false)`. Here we prefer `if (pred == false)` over `if (!pred)`.

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 2]
```diff
@@ -1707,6 +1701,24 @@ public void trimUnreferencedReaders() throws IOException {
...
+            // The checkpoint is used when opening the translog to know which files should be recovered from.
+            // We now update the checkpoint to ignore the file we are going to remove.
+            // Note that there is a provision in recoverFromFiles to allow for the case where we synced the checkpoint
+            // but crashed before we could delete the file.
+            if (!toDeleteReaderList.isEmpty()) {
+                sync();
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java]: We can skip `sync` if the `getLastSyncedCheckpoint().minTranslogGeneration >= minReferencedGen`. Can you benchmark if this optimization improves the translog performance?

**Review by dnhatn — COMMENTED** [review]
@dengweisysu This change looks great. Thank you for working on this. I left a small ask and a suggestion.

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java (hunk 2 above)
Comment (dengweisysu) [reply server/src/main/java/org/elasticsearch/index/translog/Translog.java]: Sync operation has been move out of writeLock，so  checking "getLastSyncedCheckpoint().minTranslogGeneration >= minReferencedGen"  before  sync show no significant effect.

**Comment by dnhatn** [comment]
@elasticmachine update branch

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks @dengweisysu.

**Comment by dengweisysu** [comment]
@dnhatn In discuss in last PR, you mention that we don't use the translog for recovery anymore in the future. Can you tell a little more about the new design without translog for recovery?

**Comment by dengweisysu** [comment]
@dnhatn In discuss in last PR, you mention that we don't use the translog for recovery anymore in the future. Can you tell a little more about the new design without translog for recovery? Thanks

**Comment by dnhatn** [comment]
@dengweisysu We still use translog for the local recovery (i.e., recovering the primary shard). With #45136, we start using Lucene documents in peer recovery (i.e., recovering replica shards) instead of translog.

**Comment by dnhatn** [comment]
Hi @dengweisysu,

Unfortunately, with this change, we [won't ](https://github.com/elastic/elasticsearch/issues/46267)clean up all unreferenced generations when reopening. We assume that there's [at most one](https://github.com/elastic/elasticsearch/blob/a4ed7b1ca102267ecdda7c788713675ed3958f3f/server/src/main/java/org/elasticsearch/index/translog/Translog.java#L249-L252) unreferenced generation when reopening translog. The previous implementation guarantees this assumption by syncing translog [every time](https://github.com/elastic/elasticsearch/blob/41d3eb31946a1f6f894fe1c46f57260e2195704e/server/src/main/java/org/elasticsearch/index/translog/Translog.java#L1700) after we [remove](https://github.com/elastic/elasticsearch/blob/41d3eb31946a1f6f894fe1c46f57260e2195704e/server/src/main/java/org/elasticsearch/index/translog/Translog.java#L1692) a translog reader (from the `readers` list). This change, however, only syncs translog once after we have removed all unreferenced readers (can be more than one); hence can break the assumption if this sync fails. For more detail, please see the test failure at https://github.com/elastic/elasticsearch/issues/46267.

I have reverted this commit in https://github.com/elastic/elasticsearch/commit/41d3eb31946a1f6f894fe1c46f57260e2195704e. We need to discuss with the team on how to adopt this change. I see some options for it:

1. Keep this change as is, and do not rely on the assumption. When opening translog, we need to iterate in reverse to remove unreferenced generations until neither the translog file nor the checkpoint found.

2. Remove unreferenced translog files one by one in `trimUnreferencedReaders` but syncing translog without writeLock.

@ywelsch @henningandersen WDYT?

**Comment by dengweisysu** [comment]
@dnhatn is the problem solved? the final option havn't been decided yet?

**Comment by henningandersen** [comment]
I prefer option 2, to avoid having too much cleanup to do when starting and also seems like the safer approach.

It will complicate the trimming a bit, since we still want to ensure only one thread trims at a time. You should be able to downgrade the writeLock to a readLock before sync'ing, but will have to release it in order to pick the next file. This could be susceptible to a race with other threads adding new generations, so could be worth terminating the loop once we get to the generation that was current when we began trimming.

**Comment by dnhatn** [comment]
> I prefer option 2, to avoid having too much cleanup to do when starting and also seems like the safer approach.

+1.

@dengweisysu Would you mind picking up this again? Thank you.

**Comment by dengweisysu** [comment]
@dnhatn @henningandersen Using option 1, triming really become a little complicate. I show the code below ( to show clear I temporary remove catch and assert code
企业微信截图_20190911152208
).
Is this what we need ? 
Actually I perfer option 1 because  it is robust and more clear.

**Comment by dengweisysu** [comment]
@dnhatn @henningandersen Using option 2, triming really become a little complicate. I show the code below ( to show clear I temporary remove catch and assert code)
企业微信截图_20190911152208
Is this what we need ?
Actually I perfer option 1 because it is robust and more clear.

**Comment by henningandersen** [comment]
@dengweisysu , yes, it would be something along the lines of this, though I have a few comments for it:

1. I think it will not work, since AFAIK, holding a readLock while acquiring the writeLock is not supported by `ReentrantReadWriteLock`. However, the opposite will work, allowing you to acquire the writeLock first and then acquire the readLock before releasing the writeLock, ensuring another thread cannot start trimming too.
2. the minReferencedGen should be found on the first iteration only. This ensures termination of the loop, since otherwise you could in theory be racing with other thread(s) here.

The problem with solution 1 is that while it makes this part of the code more clear, it complicates the cleanup of files, in that we can now have more than one outstanding file to cleanup. While likely solvable, it seems like solution 2 is more self-contained and easily understood, since it only concerns this one method.

**Comment by dengweisysu** [comment]
@henningandersen thanks for your comments, The code is modified with your suggest and show below ( to show clear I temporary remove catch and assert code), but  I think acquire readLock before releasing the writeLock is a litte dangerous.


To make two option more clear for deciding, I show the option 1 as below:

@dnhatn @henningandersen In my opinion, option 1 is clear because it's target is clear (delete unreferenced file) and unneccessay to keep code  in lock carefully as option 2. The assumption that there's at most one unreferenced generation when reopening translog is not so obviously that making "getFilename(minGenerationToRecoverFrom - 1)" a little strange when deleting unreferenced files . And finally failure of deleting file happens really rarely.

**Comment by dengweisysu** [comment]
@henningandersen @dnhatn Do you still perfer option 2? I don't think option 2 is more self-contained because both of them will leave unreferenced translog file which should be cleaned in another method.

**Comment by henningandersen** [comment]
@dengweisysu thanks for continuing to work on this and for the reminder. Discussing this on another channel, we wanted to bring up a 3rd solution. Similar to the work you did on #45765, could it be that doing a preemptive `sync()` before acquiring the writeLock in the original code would give a similar speed up? If so that would be the simplest solution of all.

I hope that you will benchmark that solution too so that we can compare it to original and solution 1.

Also, for completeness, it could be good to also benchmark solution 2 to see if there are significant differences between solution 1 and 2 performance-wise.

**Comment by dengweisysu** [comment]
image
the benchmark show the difference between the original code and the code doing sync() before write lock. it  has the similar effect like  #45765 . Compared with solution 2 ,  I think solution 3 may be better.

**Comment by dnhatn** [comment]
@dengweisysu Great, can you open a PR with solution 3? Thank you.

**Comment by dengweisysu** [comment]
@dnhatn I will have a travel to Paris from tomorrow to October 8.  I can do it  after I come back to China. Or you can open it If it is convenient to you ^_^

**Pull request merged**